# <center>Đồ án 3: Linear Regression</center>

# Thông tin sinh viên

- Họ và tên: Nguyễn Quốc Huy
- MSSV: 21127511
- Lớp: 21CLC02

# Import

In [50]:
import pandas as pd
import numpy as np

# Import thêm dữ thư viện nếu cần

# Đọc dữ liệu

In [51]:
# Đọc dữ liệu bằng pandas
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)
X_train = train.iloc[:, :-1]    # Dataframe (chứa các đặc trưng huấn luyện)
y_train = train.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)

X_test = test.iloc[:, :-1]      # Dataframe (chứa các đặc trưng kiểm tra)
y_test = test.iloc[:, -1]       # Series    (chứa 1 giá trị mục tiêu kiểm tra)

# Sinh viên có thể sử dụng các khác nếu cần
X_train_a = train.iloc[:, :11]    # Dataframe (chứa các đặc trưng huấn luyện)  

X_test_a = test.iloc[:, :11]      # Dataframe (chứa các đặc trưng kiểm tra)




# Cài đặt hàm

In [52]:
# Cài đặt các hàm cần thiết ở đây

class OLSLinearRegression:
    def fit(self, X, y):
        X_pinv = np.linalg.inv(X.T @ X) @ X.T    # np.linalg.pinv(X)
        self.w = X_pinv @ y

        return self


    def get_params(self):
        return self.w


    def predict(self, X):
        return np.sum(self.w.ravel() * X, axis=1)

def mae(y, y_hat):
    return np.mean(np.abs(y.ravel() - y_hat.ravel()))

def FrameToNumpy(X_train, y_train, X_test, y_test):
    return X_train.to_numpy(), y_train.to_numpy().reshape(-1,1), X_test.to_numpy(), y_test.to_numpy().reshape(-1,1)


# Yêu cầu 1a: Sử dụng toàn bộ 11 đặc trưng đầu tiên `Gender`, `10percentage`, `12percentage`, `CollegeTier`, `Degree`, `collegeGPA`, `CollegeCityTier`, `English`, `Logical`, `Quant`, `Domain` (2 điểm) 

In [53]:
# Phần code cho yêu cầu 1a

X_train_a, y_train_a, X_test_a, y_test_a = FrameToNumpy(X_train_a, y_train_a, X_test_a, y_test_a)
linearRegression = OLSLinearRegression().fit(X_train_a, y_train_a)
y_test_predict = linearRegression.predict(X_test_a)
w_array = linearRegression.get_params()

# Print results
print('w values----------------')
for i in range (0,len(w_array)):
    print(f'w{i+1}: {w_array[i][0]}')


w values----------------
w1: -22756.51282098319
w2: 804.503155673728
w3: 1294.654565252616
w4: -91781.89753057044
w5: 23182.38867915955
w6: 1437.5486723058623
w7: -8570.661984855595
w8: 147.85829929013514
w9: 152.88847647988683
w10: 117.22184593042724
w11: 34552.28622082747


In [65]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra
Res_MAE = mae(y_test_a, y_test_predict)
print('MAE:', Res_MAE)

MAE: 104863.77754033018


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = -22756.513 * Gender + 804.503 * 10percentage + 1294.655 * 12percentage - 91781.898 * CollegeTier + 23182.389 * Degree + 1437.549 * collegeGPA - 8570.662 * CollegeCityTier + 147.858 * English + 152.888 * Logical + 117.222 * Quant + 34552.286 * Domain$$

# Yêu cầu 1b: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng tính cách với các đặc trưng tính cách gồm `conscientiousness`, `agreeableness`, `extraversion`, `nueroticism`, `openess_to_experience`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [55]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

In [56]:
# Huấn luyện lại mô hình best_personality_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện

In [57]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_personality_feature_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$

# Yêu cầu 1c: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng `English`, `Logical`, `Quant`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [58]:
# Phần code cho yêu cầu 1c
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

In [59]:
# Huấn luyện lại mô hình best_skill_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện

In [60]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_skill_feature_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$

# Yêu cầu 1d: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

## Tìm mô hình

In [61]:
# Trình bày các phần tìm ra mô hình

## Thử nghiệm, so sánh các mô hình

In [62]:
# Phần code cho yêu cầu 1d
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu

In [63]:
# Huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện

In [64]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$